File: app/src/main/java/com/example/android/moviesremake/details/DetailAdapter.java
package com.example.android.moviesremake.details;

/**
 * Created by PepovPC on 7/16/2017.
 * Adapter, ktory zobrazi jednotlive views
 */

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.android.moviesremake.R;
import com.example.android.moviesremake.retrofit.MovieRetrofitReview;
import com.example.android.moviesremake.retrofit.MovieRetrofitTrailer;

import java.util.ArrayList;
import java.util.List;


public class DetailAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final String TAG = DetailAdapter.class.getSimpleName();
    private Context context;
    private final ForecastAdapterOnClickHandler mClickHandler;
    private static final int VIEW_TRAILERS = 0;
    private static final int VIEW_REVIEWS = 1;
    private List<MovieRetrofitReview> reviews = new ArrayList<>();
    private List<MovieRetrofitTrailer> trailers = new ArrayList<>();


    public interface ForecastAdapterOnClickHandler {
        void onClick(String trailer);
    }

    public DetailAdapter(Context context, ForecastAdapterOnClickHandler clickHandler) {
        this.context = context;
        mClickHandler = clickHandler;   // for Clicking
    }


    public class trailerViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener{

        public final TextView trailer;

        public trailerViewHolder(View itemView) {
            super(itemView);
            trailer = (TextView) itemView.findViewById(R.id.trailer);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            int adapterPosition = getAdapterPosition();
            String url = trailers.get(adapterPosition).getYoutubeSource();
            mClickHandler.onClick(url);
        }
    }

    public class reviewViewHolder extends RecyclerView.ViewHolder{

        public final TextView name;
        public final TextView text;

        public reviewViewHolder(View itemView) {
            super(itemView);
            name = (TextView) itemView.findViewById(R.id.name);
            text = (TextView) itemView.findViewById(R.id.text);
        }

    }


    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {

        LayoutInflater layoutInflater = LayoutInflater.from(viewGroup.getContext());

        if (viewType == VIEW_TRAILERS) {
            View v = layoutInflater.inflate(R.layout.trailer_item, viewGroup, false);

            return new trailerViewHolder(v);
        } else {
            View v = layoutInflater.inflate(R.layout.review_item, viewGroup, false);

            return new reviewViewHolder(v);
        }

    }


    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        int viewType = getItemViewType(position);
        switch(viewType){
            case VIEW_TRAILERS:
                ((trailerViewHolder) holder).trailer.setText(trailers.get(position).getTrailerName());
                String str = trailers.get(position).getTrailerName();
                System.out.println(str);
                break;
            case VIEW_REVIEWS:
                ((reviewViewHolder) holder).name.setText(reviews.get(position).getAutorName());
                ((reviewViewHolder) holder).text.setText(reviews.get(position).getAutorComment());

                break;
            default:
                System.out.println("Error in onBindViewHolder method");
        }


    }

    @Override
    public int getItemViewType(int position) {
        if(reviews.size() == 0 && trailers.size() != 0)
            return VIEW_TRAILERS;
        else
            return VIEW_REVIEWS;
    }


    @Override
    public int getItemCount() {
        if (reviews == null || reviews.size() == 0)
            return trailers.size();
        else
            return reviews.size();
    }


    public void setReviewData(List<MovieRetrofitReview> data) {
        if(data==null) {return;}
        reviews = data;
        notifyDataSetChanged();
    }

    public void setTrailerData(List<MovieRetrofitTrailer> data) {
        if(data==null) {return;}
        trailers = data;
        notifyDataSetChanged();
    }

}

File: app/src/main/java/com/example/android/moviesremake/details/DetailAdapter.kt
package com.example.android.moviesremake.details

/**
 * Created by PepovPC on 7/16/2017.
 * Adapter, ktory zobrazi jednotlive views
 */

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.example.android.moviesremake.R
import com.example.android.moviesremake.retrofit.model.MovieRetrofitReview
import com.example.android.moviesremake.retrofit.model.MovieRetrofitTrailer
import java.util.*


class DetailAdapter(private val context: Context, private val mClickHandler: ForecastAdapterOnClickHandler)// for Clicking
    : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var reviews: List<MovieRetrofitReview>? = ArrayList()
    private var trailers: List<MovieRetrofitTrailer>? = ArrayList()


    interface ForecastAdapterOnClickHandler {
        fun onClick(trailer: String?)
    }


    inner class trailerViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        val trailer: TextView

        init {
            trailer = itemView.findViewById(R.id.trailer) as TextView
            itemView.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            val adapterPosition = adapterPosition
            val url = trailers!![adapterPosition].youtubeSource
            mClickHandler.onClick(url)
        }
    }

    inner class reviewViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val name: TextView
        val text: TextView

        init {
            name = itemView.findViewById(R.id.name) as TextView
            text = itemView.findViewById(R.id.text) as TextView
        }

    }


    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): RecyclerView.ViewHolder {

        val layoutInflater = LayoutInflater.from(viewGroup.context)

        if (viewType == VIEW_TRAILERS) {
            val v = layoutInflater.inflate(R.layout.trailer_item, viewGroup, false)

            return trailerViewHolder(v)
        } else {
            val v = layoutInflater.inflate(R.layout.review_item, viewGroup, false)

            return reviewViewHolder(v)
        }

    }


    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val viewType = getItemViewType(position)
        when (viewType) {
            VIEW_TRAILERS -> {
                (holder as trailerViewHolder).trailer.text = trailers!![position].trailerName
                val str = trailers!![position].trailerName
                println(str)
            }
            VIEW_REVIEWS -> {
                (holder as reviewViewHolder).name.text = reviews!![position].autorName
                holder.text.text = reviews!![position].getAutorComment()
            }
            else -> println("Error in onBindViewHolder method")
        }


    }

    override fun getItemViewType(position: Int): Int {
        return if (reviews!!.size == 0 && trailers!!.size != 0)
            VIEW_TRAILERS
        else
            VIEW_REVIEWS
    }


    override fun getItemCount(): Int {
        return if (reviews == null || reviews!!.size == 0)
            trailers!!.size
        else
            reviews!!.size
    }


    fun setReviewData(data: List<MovieRetrofitReview>?) {
        if (data == null) {
            return
        }
        reviews = data
        notifyDataSetChanged()
    }

    fun setTrailerData(data: List<MovieRetrofitTrailer>?) {
        if (data == null) {
            return
        }
        trailers = data
        notifyDataSetChanged()
    }

    companion object {

        private val TAG = DetailAdapter::class.java.simpleName
        private val VIEW_TRAILERS = 0
        private val VIEW_REVIEWS = 1
    }

}